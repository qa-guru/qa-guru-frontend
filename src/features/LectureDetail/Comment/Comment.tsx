import React, { useEffect, useState } from "react";
import {
  Box,
  CircularProgress,
  IconButton,
  Paper,
  Stack,
  Typography,
} from "@mui/material";
import InfiniteScroll from "react-infinite-scroll-component";
import { IComment } from "./Comment.types";
import UpdateComment from "./UpdateComment";
import SendComment from "./SendComment";
import TextSerialization from "../../../shared/Serializers/TextSerialization";
import { ReactComponent as Edit } from "../../../assets/icons/button-edit.svg";
import { grey, primary } from "../../../theme/colors";
import Profile from "../Profile/Profile";

const style = {
  avatar: {
    width: 40,
    height: 40,
  },
  container: {
    backgroundColor: grey.secondary,
    borderRadius: "12px",
    padding: { xs: "10px", md: "15px" },
  },
  loadMoreBtn: { color: primary.main, margin: "0 auto" },
};

const Comment: React.FC<IComment> = (props) => {
  const { dataCommentsHomeWorkByHomeWork, dataUserId, fetchMore, id } = props;
  const [selectedIndex, setSelectedIndex] = useState<number>(-1);
  const [loading, setLoading] = useState<boolean>(false);
  const [hasMoreComments, setHasMoreComments] = useState(true);
  const { totalElements, items, offset } =
    dataCommentsHomeWorkByHomeWork?.commentsHomeWorkByHomeWork! || {};

  const handleLoadMore = () => {
    setLoading(true);
    fetchMore({
      variables: {
        offset: items?.length,
      },
      updateQuery: (
        prev: { commentsHomeWorkByHomeWork: { items: any } },
        { fetchMoreResult }: any
      ) => {
        if (!fetchMoreResult) return prev;
        return {
          commentsHomeWorkByHomeWork: {
            ...fetchMoreResult.commentsHomeWorkByHomeWork,
            items: [
              ...prev.commentsHomeWorkByHomeWork.items,
              ...fetchMoreResult.commentsHomeWorkByHomeWork.items,
            ],
          },
        };
      },
    }).then(() => setLoading(false));
  };

  useEffect(() => {
    if (items?.length! >= totalElements) {
      setHasMoreComments(false);
    }
  }, [items]);

  return (
    <>
      <SendComment id={id!} />
      <Stack spacing={1} direction="row">
        <Typography variant="h5">Комментарии</Typography>
        <Typography variant="h5">({totalElements})</Typography>
      </Stack>
      <InfiniteScroll
        dataLength={items?.length!}
        next={handleLoadMore}
        hasMore={hasMoreComments}
        loader={
          <Box mt="10px" display="flex" justifyContent="center">
            <CircularProgress size={25} />
          </Box>
        }
        style={{ overflow: "visible" }}
        scrollableTarget="scroll-container"
      >
        <Stack mt="5px" spacing={2}>
          {items?.map((item, index) => {
            const isSelected = index === selectedIndex;
            const { creator, content, creationDate, id } = item!;
            const editAccess = dataUserId?.user?.id === creator?.id;

            return (
              <Paper key={index} sx={style.container}>
                <Stack
                  alignItems={{ xs: "flex-start", md: "center" }}
                  direction="row"
                  justifyContent="space-between"
                  spacing={1}
                >
                  <Box width="100%">
                    <Profile
                      firstName={creator?.firstName!}
                      lastName={creator?.lastName!}
                      date={creationDate!}
                    />

                    <Box mt="7px">
                      {isSelected ? (
                        <UpdateComment
                          content={content!}
                          setSelectedIndex={setSelectedIndex}
                          id={id!}
                        />
                      ) : (
                        <TextSerialization text={content!} />
                      )}
                    </Box>
                  </Box>

                  {!isSelected && editAccess && (
                    <IconButton onClick={() => setSelectedIndex(index)}>
                      <Edit />
                    </IconButton>
                  )}
                </Stack>
              </Paper>
            );
          })}
        </Stack>
      </InfiniteScroll>
    </>
  );
};

export default Comment;
