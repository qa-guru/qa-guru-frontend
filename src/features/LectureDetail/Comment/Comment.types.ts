import { ReactNode } from "react";
import {
  CommentsHomeWorkByHomeWorkQuery,
  UserQuery,
} from "../../../api/graphql/generated/graphql";

export interface IComment {
  dataCommentsHomeWorkByHomeWork?: CommentsHomeWorkByHomeWorkQuery;
  dataUserId?: UserQuery;
  id?: string;
  fetchMore?: any;
}

export interface ICommentContainer {
  id: string;
  children: ReactNode;
}
