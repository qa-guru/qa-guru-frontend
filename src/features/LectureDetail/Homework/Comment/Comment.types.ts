import {
  CommentsHomeWorkByHomeWorkQuery,
  UserQuery,
} from "../../../../api/graphql/generated/graphql";

export interface IComment {
  dataCommentsHomeWorkByHomeWork: CommentsHomeWorkByHomeWorkQuery;
  dataUser: UserQuery;
  id: string;
  fetchMore: any;
}

export interface ICommentContainer {
  id: string;
}
