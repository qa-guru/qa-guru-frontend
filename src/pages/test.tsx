import { FC } from "react";
import { useParams } from "react-router-dom";

import TestContainer from "features/test/containers/test-container";

const TestPage: FC = () => {
  const { testId, trainingId, lectureId } = useParams<{
    testId: string;
    trainingId: string;
    lectureId: string;
  }>();

  // Проверяем, что все параметры присутствуют
  if (!testId || !trainingId || !lectureId) {
    return <div>Ошибка: отсутствуют необходимые параметры</div>;
  }

  return (
    <TestContainer
      testId={testId}
      trainingId={trainingId}
      lectureId={lectureId}
    />
  );
};

export default TestPage;
