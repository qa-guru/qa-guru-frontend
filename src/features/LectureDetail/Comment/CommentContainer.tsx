import React, { useState } from "react";
import Comment from "./Comment";
import { ICommentContainer } from "./Comment.types";
import { useCommentsHomeWorkByHomeWorkQuery } from "../../../api/graphql/homeworkComment/commentsHomeWorkByHomeWork";
import NoDataErrorMessage from "../../../shared/NoDataErrorMessage";
import SkeletonComment from "../../../shared/Skeletons/SkeletonComment";
import {
  CommentHomeWorkSortField,
  InputMaybe,
  Order,
} from "../../../api/graphql/generated/graphql";
import { useUserIdQuery } from "../../../api/graphql/user/userId";

const CommentContainer: React.FC<ICommentContainer> = (props) => {
  const { id, children } = props;

  const { data: dataUserId, loading: loadingUserId } = useUserIdQuery();
  const fieldSortComments =
    "CREATION_DATE" as InputMaybe<CommentHomeWorkSortField>;
  const fieldOrderComments = "DESC" as InputMaybe<Order>;

  const {
    loading: loadingCommentsHomeWorkByHomeWork,
    data: dataCommentsHomeWorkByHomeWork,
    fetchMore,
  } = useCommentsHomeWorkByHomeWorkQuery({
    variables: {
      offset: 0,
      limit: 3,
      homeWorkId: id,
      sort: {
        field: fieldSortComments,
        order: fieldOrderComments,
      },
    },
  });

  if (loadingCommentsHomeWorkByHomeWork || loadingUserId)
    return <SkeletonComment />;

  if (!dataUserId || !dataCommentsHomeWorkByHomeWork)
    return <NoDataErrorMessage />;

  return (
    <>
      {React.cloneElement(children as React.ReactElement<any>, {
        id,
        dataUserId,
        dataCommentsHomeWorkByHomeWork,
        fetchMore,
      })}
    </>
  );
};

export default CommentContainer;
