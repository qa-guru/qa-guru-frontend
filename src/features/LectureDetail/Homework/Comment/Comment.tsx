import React, { useEffect, useState } from "react";
import { Box, IconButton, Paper, Stack, Typography } from "@mui/material";
import Avatar from "@mui/material/Avatar";
import { format, parseISO } from "date-fns";
import { LoadingButton } from "@mui/lab";
import { IComment } from "./Comment.types";
import UpdateComment from "./UpdateComment";
import SendComment from "./SendComment";
import TextSerialization from "../../../../shared/Serializers/TextSerialization";
import { ReactComponent as Edit } from "../../../../assets/icons/button-edit.svg";
import { grey, primary } from "../../../../theme/colors";

const style = {
  avatar: {
    width: 40,
    height: 40,
  },
  container: {
    backgroundColor: grey.secondary,
    borderRadius: "12px",
    padding: { xs: "10px", md: "15px" },
  },
  loadMoreBtn: { color: primary.main, margin: "0 auto" },
};

const Comment: React.FC<IComment> = (props) => {
  const { dataCommentsHomeWorkByHomeWork, dataUser, fetchMore, id } = props;
  const [selectedIndex, setSelectedIndex] = useState<number>(-1);
  const [loading, setLoading] = useState<boolean>(false);
  const [hasMoreComments, setHasMoreComments] = useState(true);
  const { totalElements, items, offset } =
    dataCommentsHomeWorkByHomeWork.commentsHomeWorkByHomeWork! || {};
  const { id: idUser } = dataUser.user!;

  const handleLoadMore = () => {
    setLoading(true);
    fetchMore({
      variables: {
        offset: items?.length,
      },
      updateQuery: (
        prev: { commentsHomeWorkByHomeWork: { items: any } },
        { fetchMoreResult }: any
      ) => {
        if (!fetchMoreResult) return prev;
        return {
          commentsHomeWorkByHomeWork: {
            ...fetchMoreResult.commentsHomeWorkByHomeWork,
            items: [
              ...prev.commentsHomeWorkByHomeWork.items,
              ...fetchMoreResult.commentsHomeWorkByHomeWork.items,
            ],
          },
        };
      },
    }).then(() => setLoading(false));
  };

  useEffect(() => {
    if (items?.length! >= totalElements) {
      setHasMoreComments(false);
    }
  }, [items]);

  return (
    <Box mt="20px" p="0 15px">
      <Stack spacing={1} direction="row">
        <Typography variant="h5">Комментарии</Typography>
        <Typography variant="h5">({totalElements})</Typography>
      </Stack>
      <Stack mt="5px" spacing={2}>
        {items?.map((item, index) => {
          const isSelected = index === selectedIndex;
          const { creator, content, creationDate, id } = item!;
          const editAccess = idUser === creator?.id;

          return (
            <Paper key={index} sx={style.container}>
              <Stack
                alignItems={{ xs: "flex-start", md: "center" }}
                direction="row"
                justifyContent="space-between"
                spacing={1}
              >
                <Box width="100%">
                  <Stack spacing={1.7} direction="row" alignItems="center">
                    <Avatar
                      sx={style.avatar}
                      alt="Remy Sharp"
                      src="/static/images/avatar/1.jpg"
                    />
                    <Box>
                      <Typography variant="subtitle1">
                        {creator?.firstName} {creator?.lastName}
                      </Typography>
                      <Stack direction="row" justifyContent="space-between">
                        <Typography variant="subtitle2">
                          {format(
                            parseISO(creationDate!),
                            "dd.MM.yyyy '|' HH:mm"
                          )}
                        </Typography>
                      </Stack>
                    </Box>
                  </Stack>

                  <Box mt="7px">
                    {isSelected ? (
                      <UpdateComment
                        content={content!}
                        setSelectedIndex={setSelectedIndex}
                        id={id!}
                      />
                    ) : (
                      <TextSerialization text={content!} />
                    )}
                  </Box>
                </Box>

                {!isSelected && editAccess && (
                  <IconButton onClick={() => setSelectedIndex(index)}>
                    <Edit />
                  </IconButton>
                )}
              </Stack>
            </Paper>
          );
        })}
      </Stack>
      {hasMoreComments && (
        <Stack mt="15px">
          <LoadingButton
            loading={loading}
            onClick={handleLoadMore}
            sx={style.loadMoreBtn}
            variant="outlined"
          >
            Загрузить еще
          </LoadingButton>
        </Stack>
      )}
      <SendComment id={id!} />
    </Box>
  );
};

export default Comment;
