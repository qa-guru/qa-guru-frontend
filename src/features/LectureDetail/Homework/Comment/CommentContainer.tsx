import React from "react";
import Comment from "./Comment";
import { ICommentContainer } from "./Comment.types";
import { useCommentsHomeWorkByHomeWorkQuery } from "../../../../api/graphql/homeworkComment/commentsHomeWorkByHomeWork";
import NoDataErrorMessage from "../../../../shared/NoDataErrorMessage";
import SkeletonComment from "../../../../shared/Skeletons/SkeletonComment";
import { useUserQuery } from "../../../../api/graphql/user/user";
import {
  CommentHomeWorkSortField,
  InputMaybe,
  Order,
} from "../../../../api/graphql/generated/graphql";

const CommentContainer: React.FC<ICommentContainer> = (props) => {
  const { id } = props;
  const { data: dataUser, loading: loadingUser } = useUserQuery();
  const fieldSortComments =
    "CREATION_DATE" as InputMaybe<CommentHomeWorkSortField>;
  const fieldOrderComments = "DESC" as InputMaybe<Order>;

  const {
    loading: loadingCommentsHomeWorkByHomeWork,
    data: dataCommentsHomeWorkByHomeWork,
    fetchMore,
  } = useCommentsHomeWorkByHomeWorkQuery({
    variables: {
      offset: 0,
      limit: 3,
      homeWorkId: id,
      sort: {
        field: fieldSortComments,
        order: fieldOrderComments,
      },
    },
  });

  if (loadingCommentsHomeWorkByHomeWork || loadingUser)
    return <SkeletonComment />;
  if (!dataUser || !dataCommentsHomeWorkByHomeWork)
    return <NoDataErrorMessage />;

  return (
    <Comment
      id={id}
      dataUser={dataUser}
      dataCommentsHomeWorkByHomeWork={dataCommentsHomeWorkByHomeWork}
      fetchMore={fetchMore}
    />
  );
};

export default CommentContainer;
