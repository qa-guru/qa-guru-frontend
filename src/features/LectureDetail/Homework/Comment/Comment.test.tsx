import { MemoryRouter } from "react-router-dom";
import Comment from "./Comment";
import { render } from "../../../../test/utilsTest";
import { mockDataUser } from "../../../../shared/mocks/dataUser.mock";
import { mockDataCommentsHomeWorkByHomeWork } from "../../../../shared/mocks/dataCommentsHomeWorkByHomeWork.mock";

const mockFetchMoreCommentsHomeWorkByHomeWork: any = () => {};

describe("Comment", () => {
  it("the component is render", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <Comment
          id="string"
          dataUser={mockDataUser}
          dataCommentsHomeWorkByHomeWork={mockDataCommentsHomeWorkByHomeWork}
          fetchMore={mockFetchMoreCommentsHomeWorkByHomeWork}
        />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
