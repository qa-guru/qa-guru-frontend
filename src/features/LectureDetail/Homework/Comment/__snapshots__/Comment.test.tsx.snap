// Vitest Snapshot v1

exports[`Comment > the component is render 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-a4a06o"
  >
    <div
      class="MuiStack-root css-w4z10b-MuiStack-root"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Комментарии
      </h5>
      <h5
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        ()
      </h5>
    </div>
    <div
      class="MuiStack-root css-1w1iguc-MuiStack-root"
    />
    <div
      class="MuiStack-root css-1sxz4pq-MuiStack-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiLoadingButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-181t86v-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
        id=":r0:"
        tabindex="0"
        type="button"
      >
        Загрузить еще
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <form>
      <h5
        class="MuiTypography-root MuiTypography-h5 css-6nzxz6-MuiTypography-root"
      >
        Добавить комментарий
      </h5>
      <div
        class="MuiBox-root css-8atqhb"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <textarea
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r1:"
                placeholder="Текст ответа"
                style="height: 0px; overflow: hidden;"
              />
              <textarea
                aria-hidden="true"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                readonly=""
                style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                tabindex="-1"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiStack-root css-1qvzoxq-MuiStack-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-17d8meg-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
          id=":r2:"
          tabindex="0"
          type="button"
        >
          Отправить
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
